### PR TITLE
Backport use expression context for conditional visibility

### LINKED ISSUE #37154: Forms: fix for control visibility by expression needs to be backported to 3.10
**Describe the bug**
Tabs in attribute forms can be shown/hidden based on an expression.

image

While this works fine in QGIS master, it fails in QGIS 3.10.x

According to @m-kuhn the automatic backport of a fix in master to 3.10 failed and was forgotten to do it manually.

@m-kuhn can you please reference the fix in master?

### PR DESCRIPTION
**Pull request by m-kuhn**
## Description

Backports #35881
Fixes #37154

### CODE REVIEW

**Comment by m-kuhn** [comment]
Seems to go into an infinite loop with datetime edit.

@nyalldawson any idea? Could this be related to no having 983a3d8655b (random guess)?

```
#11813 0x00007fffe55544d1 in QgsDateTimeEditWrapper::dateTimeChanged(QDateTime const&) (this=0x5555567fb8d0, dateTime=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/qgsdatetimeeditwrapper.cpp:130
#11814 0x00007fffe5555d5b in QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<QDateTime const&>, void, void (QgsDateTimeEditWrapper::*)(QDateTime const&)>::call(void (QgsDateTimeEditWrapper::*)(QDateTime const&), QgsDateTimeEditWrapper*, void**) (f=(void (QgsDateTimeEditWrapper::*)(QgsDateTimeEditWrapper * const, const QDateTime &)) 0x7fffe55542f8 <QgsDateTimeEditWrapper::dateTimeChanged(QDateTime const&)>, o=0x5555567fb8d0, arg=0x7fffff974510) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:152
#11815 0x00007fffe5555cde in QtPrivate::FunctionPointer<void (QgsDateTimeEditWrapper::*)(QDateTime const&)>::call<QtPrivate::List<QDateTime const&>, void>(void (QgsDateTimeEditWrapper::*)(QDateTime const&), QgsDateTimeEditWrapper*, void**) (f=(void (QgsDateTimeEditWrapper::*)(QgsDateTimeEditWrapper * const, const QDateTime &)) 0x7fffe55542f8 <QgsDateTimeEditWrapper::dateTimeChanged(QDateTime const&)>, o=0x5555567fb8d0, arg=0x7fffff974510) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:185
#11816 0x00007fffe5555c48 in QtPrivate::QSlotObject<void (QgsDateTimeEditWrapper::*)(QDateTime const&), QtPrivate::List<QDateTime const&>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x5555566ae230, r=0x5555567fb8d0, a=0x7fffff974510, ret=0x0) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:414
#11817 0x00007fffea104ac0 in QMetaObject::activate(QObject*, int, int, void**) () at /lib64/libQt5Core.so.5
#11818 0x00007fffe5a00d55 in QgsDateTimeEdit::valueChanged(QDateTime const&) (this=0x55555661c860, _t1=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/build/src/gui/editorwidgets/moc_qgsdatetimeedit.cpp:183
#11819 0x00007fffe554a6a5 in QgsDateTimeEdit::changed(QDateTime const&) (this=0x55555661c860, dateTime=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/qgsdatetimeedit.cpp:216
#11820 0x00007fffe554b676 in QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<QDateTime const&>, void, void (QgsDateTimeEdit::*)(QDateTime const&)>::call(void (QgsDateTimeEdit::*)(QDateTime const&), QgsDateTimeEdit*, void**) (f=(void (QgsDateTimeEdit::*)(QgsDateTimeEdit * const, const QDateTime &)) 0x7fffe554a54c <QgsDateTimeEdit::changed(QDateTime const&)>, o=0x55555661c860, arg=0x7fffff974720) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:152
#11821 0x00007fffe554b575 in QtPrivate::FunctionPointer<void (QgsDateTimeEdit::*)(QDateTime const&)>::call<QtPrivate::List<QDateTime const&>, void>(void (QgsDateTimeEdit::*)(QDateTime const&), QgsDateTimeEdit*, void**) (f=(void (QgsDateTimeEdit::*)(QgsDateTimeEdit * const, const QDateTime &)) 0x7fffe554a54c <QgsDateTimeEdit::changed(QDateTime const&)>, o=0x55555661c860, arg=0x7fffff974720) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:185
#11822 0x00007fffe554b49a in QtPrivate::QSlotObject<void (QgsDateTimeEdit::*)(QDateTime const&), QtPrivate::List<QDateTime const&>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x55555671f230, r=0x55555661c860, a=0x7fffff974720, ret=0x0) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:414
#11823 0x00007fffea104ac0 in QMetaObject::activate(QObject*, int, int, void**) () at /lib64/libQt5Core.so.5
#11824 0x00007fffde2e1596 in QDateTimeEdit::dateTimeChanged(QDateTime const&) () at /lib64/libQt5Widgets.so.5
#11825 0x00007fffde2e7154 in QDateTimeEditPrivate::emitSignals(EmitPolicy, QVariant const&) () at /lib64/libQt5Widgets.so.5
#11826 0x00007fffde373949 in QAbstractSpinBoxPrivate::setValue(QVariant const&, EmitPolicy, bool) () at /lib64/libQt5Widgets.so.5
#11827 0x00007fffde2e2dd2 in QDateTimeEdit::setDateTime(QDateTime const&) () at /lib64/libQt5Widgets.so.5
#11828 0x00007fffe554ac99 in QgsDateTimeEdit::setDateTime(QDateTime const&) (this=0x55555661c860, dateTime=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/qgsdatetimeedit.cpp:281
#11829 0x00007fffe5555790 in QgsDateTimeEditWrapper::updateValues(QVariant const&, QList<QVariant> const&) (this=0x5555567fb8d0, value=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/qgsdatetimeeditwrapper.cpp:245
#11830 0x00007fffe553ed19 in QgsEditorWidgetWrapper::setValues(QVariant const&, QList<QVariant> const&) (this=0x5555567fb8d0, value=..., additionalValues=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp:88
#11831 0x00007fffe56c9127 in QtPrivate::FunctorCall<QtPrivate::IndexesList<0, 1>, QtPrivate::List<QVariant const&, QList<QVariant> const&>, void, void (QgsEditorWidgetWrapper::*)(QVariant const&, QList<QVariant> const&)>::call(void (QgsEditorWidgetWrapper::*)(QVariant const&, QList<QVariant> const&), QgsEditorWidgetWrapper*, void**) (f=(void (QgsEditorWidgetWrapper::*)(QgsEditorWidgetWrapper * const, const QVariant &, const QList<QVariant> &)) 0x7fffe553ece0 <QgsEditorWidgetWrapper::setValues(QVariant const&, QList<QVariant> const&)>, o=0x5555567fb8d0, arg=0x7fffff974c30) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:152
#11832 0x00007fffe56c8b89 in QtPrivate::FunctionPointer<void (QgsEditorWidgetWrapper::*)(QVariant const&, QList<QVariant> const&)>::call<QtPrivate::List<QVariant const&, QList<QVariant> const&>, void>(void (QgsEditorWidgetWrapper::*)(QVariant const&, QList<QVariant> const&), QgsEditorWidgetWrapper*, void**) (f=(void (QgsEditorWidgetWrapper::*)(QgsEditorWidgetWrapper * const, const QVariant &, const QList<QVariant> &)) 0x7fffe553ece0 <QgsEditorWidgetWrapper::setValues(QVariant const&, QList<QVariant> const&)>, o=0x5555567fb8d0, arg=0x7fffff974c30) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:185
#11833 0x00007fffe56c7e00 in QtPrivate::QSlotObject<void (QgsEditorWidgetWrapper::*)(QVariant const&, QList<QVariant> const&), QtPrivate::List<QVariant const&, QList<QVariant> const&>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x555556800ab0, r=0x5555567fb8d0, a=0x7fffff974c30, ret=0x0) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:414
#11834 0x00007fffea104ac0 in QMetaObject::activate(QObject*, int, int, void**) () at /lib64/libQt5Core.so.5
#11835 0x00007fffe59ff4ad in QgsEditorWidgetWrapper::valuesChanged(QVariant const&, QList<QVariant> const&) (this=0x55555657d050, _t1=..., _t2=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/build/src/gui/editorwidgets/core/moc_qgseditorwidgetwrapper.cpp:264
#11836 0x00007fffe55544d1 in QgsDateTimeEditWrapper::dateTimeChanged(QDateTime const&) (this=0x55555657d050, dateTime=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/qgsdatetimeeditwrapper.cpp:130
#11837 0x00007fffe5555d5b in QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<QDateTime const&>, void, void (QgsDateTimeEditWrapper::*)(QDateTime const&)>::call(void (QgsDateTimeEditWrapper::*)(QDateTime const&), QgsDateTimeEditWrapper*, void**) (f=(void (QgsDateTimeEditWrapper::*)(QgsDateTimeEditWrapper * const, const QDateTime &)) 0x7fffe55542f8 <QgsDateTimeEditWrapper::dateTimeChanged(QDateTime const&)>, o=0x55555657d050, arg=0x7fffff975060) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:152
#11838 0x00007fffe5555cde in QtPrivate::FunctionPointer<void (QgsDateTimeEditWrapper::*)(QDateTime const&)>::call<QtPrivate::List<QDateTime const&>, void>(void (QgsDateTimeEditWrapper::*)(QDateTime const&), QgsDateTimeEditWrapper*, void**) (f=(void (QgsDateTimeEditWrapper::*)(QgsDateTimeEditWrapper * const, const QDateTime &)) 0x7fffe55542f8 <QgsDateTimeEditWrapper::dateTimeChanged(QDateTime const&)>, o=0x55555657d050, arg=0x7fffff975060) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:185
#11839 0x00007fffe5555c48 in QtPrivate::QSlotObject<void (QgsDateTimeEditWrapper::*)(QDateTime const&), QtPrivate::List<QDateTime const&>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x55555653af90, r=0x55555657d050, a=0x7fffff975060, ret=0x0) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:414
#11840 0x00007fffea104ac0 in QMetaObject::activate(QObject*, int, int, void**) () at /lib64/libQt5Core.so.5
#11841 0x00007fffe5a00d55 in QgsDateTimeEdit::valueChanged(QDateTime const&) (this=0x5555567ef800, _t1=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/build/src/gui/editorwidgets/moc_qgsdatetimeedit.cpp:183
#11842 0x00007fffe554a6a5 in QgsDateTimeEdit::changed(QDateTime const&) (this=0x5555567ef800, dateTime=...) at /home/mkuhn/dev/qgis/backport-35881-release-3_10/src/gui/editorwidgets/qgsdatetimeedit.cpp:216
#11843 0x00007fffe554b676 in QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<QDateTime const&>, void, void (QgsDateTimeEdit::*)(QDateTime const&)>::call(void (QgsDateTimeEdit::*)(QDateTime const&), QgsDateTimeEdit*, void**) (f=(void (QgsDateTimeEdit::*)(QgsDateTimeEdit * const, const QDateTime &)) 0x7fffe554a54c <QgsDateTimeEdit::changed(QDateTime const&)>, o=0x5555567ef800, arg=0x7fffff975270) at /usr/include/qt5/QtCore/qobjectdefs_impl.h:152

```

**Comment by nyalldawson** [comment]
@m-kuhn 
>  Could this be related to no having 983a3d8 (random guess)?

Could be... I honestly can't recall that particular commit, but it definitely seems like there's an unnecessary signal being thrown around here.

File: src/gui/editorwidgets/qgsdatetimeedit.cpp [hunk 1]
```diff
@@ -192,8 +192,12 @@ void QgsDateTimeEdit::showEvent( QShowEvent *event )
 
 void QgsDateTimeEdit::changed( const QDateTime &dateTime )
 {
-  mIsEmpty = false;
   bool isNull = dateTime.isNull();
+
+  if ( mIsNull == isNull && QgsDateTimeEdit::dateTime() == dateTime )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsdatetimeedit.cpp:197]: @elpaso this is more or less along the lines of https://github.com/qgis/QGIS/pull/30100 but on a different level.

1. does it look good to you?
2. should we remove the (now redundant i think) [outer](https://github.com/qgis/QGIS/pull/30100/files#diff-570a1ee0ae596d03e7fda06c5f47eafeR66) [checks](https://github.com/qgis/QGIS/pull/30100/files#diff-570a1ee0ae596d03e7fda06c5f47eafeR260)?

Comment (elpaso) [reply src/gui/editorwidgets/qgsdatetimeedit.cpp:197]: I don't remember all the details of my PR but looking at the code I think you can remove the outer checks.

Comment (elpaso) [reply src/gui/editorwidgets/qgsdatetimeedit.cpp:197]: ... and yes: LGTM

**Comment by m-kuhn** [comment]
Thank you @elpaso 
If it's ok like this, can you approve?

**Review by elpaso — APPROVED** [review]

**Pull request merged**